pkg/auth/auth.go

```diff
@@ -0,0 +1,370 @@
+	// SetVerifier sets the code verifier.
+	SetVerifier(verifier *oidc.S256Verifier)
+	// Start starts a local server for handling authentication processing.
+	Start() error
+}
+
+// Server is a local server responsible for authentication processing.
+type Server struct {
+	// APIEndpoint is the API endpoint.
+	APIEndpoint string
+	// AccountEndpoint is the accounts endpoint.
+	AccountEndpoint string
+	// HTTPClient is a HTTP client used to call the API to exchange the access token for a session token.
+	HTTPClient api.HTTPClient
+	// Result is a channel that reports the result of authorization.
+	Result chan AuthorizationResult
+	// Router is an HTTP request multiplexer.
+	Router *http.ServeMux
+	// Verifier represents an OAuth PKCE code verifier that uses the S256 challenge method.
+	Verifier *oidc.S256Verifier
+}
+
+// GetResult returns the result channel.
+func (s Server) GetResult() chan AuthorizationResult {
+	return s.Result
+}
+
+// SetAccountEndpoint sets the account endpoint.
+func (s *Server) SetAccountEndpoint(endpoint string) {
+	s.AccountEndpoint = endpoint
+}
+
+// SetAPIEndpoint sets the API endpoint.
+func (s *Server) SetAPIEndpoint(endpoint string) {
+	s.APIEndpoint = endpoint
+}
+
+// SetVerifier sets the code verifier endpoint.
+func (s *Server) SetVerifier(verifier *oidc.S256Verifier) {
+	s.Verifier = verifier
+}
+
+// Start starts a local server for handling authentication processing.
+func (s *Server) Start() error {
+	server := &http.Server{
+		Addr:         ":8080",
+		Handler:      s.Router,
+		ReadTimeout:  10 * time.Second,
+		WriteTimeout: 10 * time.Second,
+	}
+
+	err := server.ListenAndServe()
+	if err != nil {
+		return fsterr.RemediationError{
+			Inner:       fmt.Errorf("failed to start local server: %w", err),
+			Remediation: Remediation,
+		}
+	}
+	return nil
+}
+
+// HandleCallback processes the callback from the authentication service.
+func (s *Server) HandleCallback() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		authorizationCode := r.URL.Query().Get("code")
+		if authorizationCode == "" {
+			fmt.Fprint(w, "ERROR: no authorization code returned\n")
+			s.Result <- AuthorizationResult{
+				Err: fmt.Errorf("no authorization code returned"),
+			}
+			return
+		}
+
+		// Exchange the authorization code and the code verifier for a JWT.
+		// NOTE: I use the identifier `j` to avoid overlap with the `jwt` package.
+		codeVerifier := s.Verifier.Verifier()
+		j, err := GetJWT(s.AccountEndpoint, codeVerifier, authorizationCode)
+		if err != nil || j.AccessToken == "" || j.IDToken == "" {
+			fmt.Fprint(w, "ERROR: failed to exchange code for JWT\n")
+			s.Result <- AuthorizationResult{
+				Err: fmt.Errorf("failed to exchange code for JWT"),
+			}
+			return
+		}
+
+		claims, err := VerifyJWTSignature(s.AccountEndpoint, j.AccessToken)
+		if err != nil {
+			s.Result <- AuthorizationResult{
+				Err: err,
+			}
+			return
+		}
+
+		email, ok := claims["email"]
```

Comment:
Before we trust the user claims in the JWT token, we need to quickly check:
1. `azp == fastly-cli`
2. `aud == https://api.fastly.com`

Reply:
Similar to this -> https://github.com/hashicorp/cap/blob/main/jwt/docs_test.go#L17